lib/cn/cn_tree.c

```diff
@@ -1855,49 +1919,170 @@ cn_comp_cleanup(struct cn_compaction_work *w)
+    struct cn_compaction_work *w = ss->ss_work;
+    struct cndb *cndb = w->cw_tree->cndb;
+
+    if (!ss->ss_added)
+        return 0;
+
+    if (!mblks->hblk.bk_blkid) {
+        assert(mblks->kblks.n_blks == 0);
+        return 0;
+    }
+
+    err = cndb_record_txstart(cndb, 0, CNDB_INVAL_INGESTID, CNDB_INVAL_HORIZON, 1, 0, &tx);
+    if (err)
+        return err;
 
-    mutex_lock(&tree->ct_rspills_lock);
+    cn_subspill_get_kvset_meta(ss, &km);
 
-    w = list_first_entry_or_null(&tree->ct_rspills_list, typeof(*w), cw_rspill_link);
-    if (!w)
+    /* CNDB: Log kvset add records.
+     */
+    err = cndb_record_kvset_add(cndb, tx, w->cw_tree->cnid, km.km_nodeid,
+                                &km, ss->ss_kvsetid, km.km_hblk.bk_blkid,
+                                mblks->kblks.n_blks, (uint64_t *)mblks->kblks.blks,
+                                mblks->vblks.n_blks, (uint64_t *)mblks->vblks.blks, &cookie);
+    if (err)
         goto done;
 
-    /* Punt if job on head of list is not done or another thread is already committing it. */
-    if (!atomic_read(&w->cw_rspill_done) || atomic_read(&w->cw_rspill_commit_in_progress)) {
-        w = 0;
+    err = cn_mblocks_commit(w->cw_mp, 1, mblks, CN_MUT_OTHER);
+    if (err)
+        goto done;
+
+    err = kvset_open(w->cw_tree, ss->ss_kvsetid, &km, &kvset);
+    if (err)
         goto done;
-    }
 
-    /* Job on head of spill completion list is ready to be processed.
-     * - Set "commit_in_progress" status, but leave on list until commit is done.
-     * - If the node is wedged, it means an earlier job has failed, in
-     *   which case we force failure on this job to prevent out of
-     *   order completion.
-     * - If the node is not wedged, and this job has failed then it
-     *   will cause the node to be wedged, but this will be handled
-     *   later to catch downstream errors.
+    /* CNDB: Ack kvset add record.
      */
+    err = cndb_record_kvset_add_ack(cndb, tx, cookie);
+    if (err)
+        goto done;
 
-    atomic_set(&w->cw_rspill_commit_in_progress, 1);
+    cn_comp_kvset_append(w, ss->ss_node, kvset);
 
-    if (ev(tree->ct_rspills_wedged && !w->cw_err)) {
-        w->cw_err = merr(ESHUTDOWN);
-        w->cw_canceled = true;
-    }
+    blk_list_free(&ss->ss_mblks.kblks);
+    blk_list_free(&ss->ss_mblks.vblks);
 
 done:
-    mutex_unlock(&tree->ct_rspills_lock);
+    w->cw_t4_commit = get_time_ns();
+
+    if (err) {
+        cndb_record_nak(cndb, tx);
+        kvset_put_ref(kvset);
+    }
+
+    return err;
+}
+
+static merr_t
+cn_node_apply_subspill(struct subspill *ss)
+{
+    struct cn_compaction_work *w = ss->ss_work;
+    merr_t err;
+
+    err = cn_subspill_commit(ss);
+    if (err)
+        return err;
+
+    if (ss->ss_added) {
+        w->cw_output_nodev[0] = ss->ss_node;
+        w->cw_checkpoint(w);
+    }
 
-    return w;
+    return 0;
+}
+
+static merr_t
+cn_comp_spill(struct cn_compaction_work *w)
```

Comment:
Maybe add a function comment acknowledging that if the spill is interrupted by a failure, the spill will start over from the beginning and respill the same keys to the same nodes.  And that it's not a significant issue since GC will clean it up.

Reply:
Does the new spill listen to `cw_cancel`?  If a close occurs in the middle of a root spill, will it stop at whatever leaf it's on?

Reply:
Yes, it will abandon the ongoing subspill and exit.